server/repository/src/db_diesel/number_row.rs

```diff
@@ -167,14 +171,13 @@ impl<'a> NumberRowRepository<'a> {
                     .bind::<Text, _>(store_id)
                     .bind::<Text, _>(r#type.to_string());
 
-                match insert_query.get_result::<NextNumber>(&self.connection.connection) {
+                let mut guard = self.connection.lock();
+                match insert_query.get_result::<NextNumber>(guard.connection()) {
```

Comment:
This is where I got a deadlock when locking in the `get_result` and later in the Err(NotFound) branch...

Reply:
Interesting! I wonder if there's some risks we need to be considering around this locking?

Reply:
This is a bit of an odd case. Usually you just do something like:
`let result = query.get_result::<...>(self.connection.lock().connection())`
and then the lock is released as expected. But good to keep it in mind!

Reply:
I hope we would have some time to make it a little bit safer, when we move to async, as per the https://github.com/msupply-foundation/open-msupply/issues/78, which would make any locking happen within closure, i think compiler would complain if you try to double lock it.

Maybe for now, we have one simple rule, to just lock at the start of a repo method (and only lock once ?). Like in this example would this work: https://github.com/msupply-foundation/open-msupply/compare/update-diesel...update-diesel(Lock-once)

Reply:
Yeah for new code this would make sense. However, the most common case is probably still a single `let result = query.get_result::<...>(self.connection.lock().connection())` call and then this is IMHO good enough. I am not too worried about deadlocks in the repo methods, the `match` example was quite unusually and for other possible deadlocks it should be straight forward to detect and find them.

And as you said being able to access the raw connection only in a `con(|con| {...})` closure would solve it as well.